Prepare til wrappers for migrating off of SMALL_RECT

lhecker opened the pull request:
This commit makes the following changes to `til::point/size/rectangle`
for the following reasons:
* Rename `rectangle` into `rect`
  This will make the naming consistent with a later `small_rect` struct
  as well as the existing Win32 POINT/SIZE/RECT structs.
* Standardizes til wrappers on `int32_t` instead of `ptrdiff_t`
  Provides a consistent behavior between x86 and x64, preventing accidental
  errors on x86, as it's less rigorously tested than x64. Additionally it
  improves interop with MIDL3 which only supports fixed width integer types.
* Standardizes til wrappers on throwing `gsl::narrow_error`
  Makes the behavior of our code more consistent.
* Makes all eligible functions `constexpr`
  Because why not.
* Removes implicit constructors and conversion operators
  This is a complex and controversial topic. My reasons are: You can't Ctrl+F
  for an implicit conversion. This breaks most non-IDE engines, like the one on
  GitHub or those we have internally at MS. This is important for me as these
  implicit conversion operators aren't cost free. Narrowing integers itself,
  as well as the boundary checks that need to be done have a certain,
  fixed overhead each time. Additionally the lack of noexcept prevents
  many advanced compiler optimizations. Removing their use entirely
  drops conhost's code segment size by around ~6.5%.

## References

Preliminary work for #4015.

## PR Checklist
* [x] I work here
* [x] Tests added/passed

## Validation Steps Performed
I'm mostly relying on our unit tests here. Both OpenConsole and WT appear to work fine.

carlos-zamora commented on src/inc/til/rect.h:
```diff
@@ -0,0 +1,851 @@
+// Copyright (c) Microsoft Corporation.
```
(note for reviewers) Use a diff tool (I used text-compare.com) to compare rect.h to rectangle.h and make your life easier haha.

lhecker replied:
It's probably better in general to check these PRs out locally and compare them VS Code for instance.
BTW VS Code has this which is nice:
![image](https://user-images.githubusercontent.com/2256941/145863775-3ff5a76b-1a0b-4512-9b33-d16f76636c4e.png)

carlos-zamora commented on src/cascadia/TerminalControl/ControlInteractivity.cpp:
```diff
@@ -293,15 +293,17 @@ namespace winrt::Microsoft::Terminal::Control::implementation
+                const auto dy = pixelPosition.y - touchdownPoint.y;
+                const auto w = fontSizeInDips.width;
+                const auto distanceSquared = dx * dx + dy * dy;
+                const auto maxDistanceSquared = w * w / 16; // (w /4)^2
```
so we're using `(w/4)^2` here instead of `(std::min(w, height)/4)^2` because w is always less than height?

lhecker replied:
Yeah it just simplifies the code. I hoped that the comment above this code block explains it sufficiently.
Apart from that this is a bog-standard distance check for CPUs. It's not about speed necessarily, it's just that the `sqrt` function is weird, whacky, and comes with a whole baggage of stuff we don't need if we just square the distance instead.

carlos-zamora commented on src/cascadia/TerminalSettingsModel/ActionArgs.idl:
```diff
@@ -5,16 +5,14 @@ import "Command.idl";
 
 namespace Microsoft.Terminal.Settings.Model
 {
-    interface IActionArgs
```
this seems unrelated, but I guess it's fine? haha

lhecker replied:
Ah that's a mistake from invoking clang-format on it. I'll revert it.

carlos-zamora reviewed (commented):
53/76
Honestly, it all looks good. Most of it seems to be a result of removing the implicit conversion stuff.

lhecker commented on src/types/UiaTextRangeBase.cpp:
```diff
@@ -287,16 +287,16 @@ void UiaTextRangeBase::_expandToEnclosingUnit(TextUnit unit)
     // set us to ONE BEFORE the document end.
     // This allows us to expand properly.
-    if (bufferSize.CompareInBounds(_start, documentEnd, true) >= 0)
+    if (bufferSize.CompareInBounds(_start, documentEnd.to_win32_coord(), true) >= 0)
```
Ah I didn't realize that. I could try and replace it.

carlos-zamora replied:
Yup! Just about all of the `CompareInBounds` calls are a little overkill. They actually are the cause of a lot of `FAILFAST` issues we've had in accessibility code. We may even be able to remove the extra bool parameter off of CompareInBounds if we make this happen. That's why I get really excited as we adopt `til::point` more in our codebase haha

lhecker commented on src/host/readData.hpp:
```diff
@@ -23,6 +23,8 @@ Revision History:
 #include "../server/IWaitRoutine.h"
 #include "../server/WaitTerminationReason.h"
 
+class InputBuffer;
+
```
One of the test files had this weird comment about "oh you need to include InputBuffer.h here first. idk why but it fails compilation otherwise". I fixed that issue by properly forward-declaring `InputBuffer` there.

carlos-zamora reviewed (commented):
72/76. Mostly tests are left. I also want to re-review the til::rect changes themselves. Looking good! :)

lhecker commented on src/cascadia/WindowsTerminal/IslandWindow.cpp:
```diff
@@ -516,11 +516,12 @@ long IslandWindow::_calculateTotalSize(const bool isWidth, const long clientSize
             const auto relative = eventPoint - origin;
             // Convert to logical scaling before raising the event.
-            const auto real = relative / GetCurrentDpiScale();
+            const auto scale = GetCurrentDpiScale();
+            const til::point real{ til::math::flooring, relative.x / scale, relative.y / scale };
```
The previous code didn't specify the exact `til::math::*` rounding mode to use. This one does.
This fixes at least one bug in our code where we use `til::point / float`, which implicitly does a flooring narrow, instead of a more appropriate `til::math::round`.

lhecker commented on src/renderer/dx/CustomTextRenderer.cpp:
```diff
@@ -273,18 +273,17 @@ try
-                               drawingContext.cellSize.height };
-
     // Create rectangular block representing where the cursor can fill.
-    D2D1_RECT_F rect = til::rectangle{ til::point{ options.coordCursor } }.scale_up(glyphSize);
```
I found this code in particular so incredibly hard to understand that I changed it into simple C code.
In order to understand the existing code you have to understand that:
* `til::point` has a constructor for `COORD` (`options.coordCursor`)
* `til::rectangle` has a constructor for a point without size, which creates a rectangle at that origin of size 1x1
* `scale_up` multiplies the size of the rectangle, but not the origin
* `til::rectangle` implicitly turns into a `D2D1_RECT_F` using a conversion operator

The interaction between the second and third point are outright confusing for me and it took me more time to understand that code than I'd like to admit. I'd be fine with this code:
```cpp
D2D1_RECT_F rect = til::rectangle{ til::point{ options.coordCursor }, glyphSize };
```

But at that point I asked myself why you'd want to construct a til::rectangle anyways, just to turn it into a `D2D1_RECT_F` right after. If you'd like me to use that alternative code anyways, I'd be absolutely fine doing so of course! But personally I preferred the C-style code.

carlos-zamora reviewed (approved):
Finally finished! 🎉🥳 Looks great! Excited to see this land 🙂

Most of the changes are really just...
1. `til::rectangle` --> `til::rect`
2. implicit --> explicit conversion

So going through most of these files are pretty easy. Anything that isn't that, I've pointed out and you've done a good job of explaining your reasoning.

I like `details::extract`. It's a nice way to simplify a lot of the assignment and add some "built-in" validation.

Great work my dude 👍

lhecker commented:
I've rebased the last commit a few times so that there's only one additional commit to review since my initial one.
Apart from "addressing feedback" it also fixes compilation issues, makes the last remaining til constructor explicit (after quite some deliberation tbh) and fixes til::math, which - as it turns out - didn't throw during narrowing. Now it does.

zadjii-msft reviewed (commented):
You know, I'm not sure I'm qualified to be the person to give the second ✔️  on the `til` stuff. Most of this looks fine. The `GenerateTests.ps1` one definitely needs fixing. But I think I'd be more comfortable with @miniksa / @DHowett's approval, since they wrote most of these originally. 

I personally really _liked_ the implicit conversions, I feel like that was a part of what made `til` so magic. But if we need to get rid of them, then I get it. So long as there's still magic constructors from `anything` to the til types, then I'll be happy.

miniksa commented:
> I personally really _liked_ the implicit conversions, I feel like that was a part of what made `til` so magic. 
I agree with this sentiment.

> But if we need to get rid of them, then I get it. So long as there's still magic constructors from `anything` to the til types, then I'll be happy.
But after Leonard was explaining to me that the implicits were making a whole bunch of errors and thrashing... I can live with the change to be more explicit. I meant to eliminate bug factories with the til types, not turn them into one.

miniksa commented on src/cascadia/WinRTUtils/inc/Utils.h:
```diff
@@ -3,23 +3,6 @@
-// - scale: amount to scale rect by
-// Return Value:
-// - Rect scaled by scale
-inline winrt::Windows::Foundation::Rect ScaleRect(winrt::Windows::Foundation::Rect rect, double scale)
```
Yay!

miniksa commented on src/cascadia/WindowsTerminal/IslandWindow.cpp:
```diff
@@ -383,11 +382,10 @@ void IslandWindow::_OnGetMinMaxInfo(const WPARAM /*wParam*/, const LPARAM lParam
-    lpMinMaxInfo->ptMinTrackSize.x = _calculateTotalSize(true, minimumWidth * scale, nonClientSizeScaled.cx);
-    lpMinMaxInfo->ptMinTrackSize.y = _calculateTotalSize(false, minimumHeight * scale, nonClientSizeScaled.cy);
+    lpMinMaxInfo->ptMinTrackSize.x = _calculateTotalSize(true, minimumWidth * dpix / USER_DEFAULT_SCREEN_DPI, nonClientSizeScaled.cx);
+    lpMinMaxInfo->ptMinTrackSize.y = _calculateTotalSize(false, minimumHeight * dpix / USER_DEFAULT_SCREEN_DPI, nonClientSizeScaled.cy);
```
```suggestion
    lpMinMaxInfo->ptMinTrackSize.y = _calculateTotalSize(false, minimumHeight * dpiy / USER_DEFAULT_SCREEN_DPI, nonClientSizeScaled.cy);

```

It's not going to be different but may as well make it look like it is sensible.

miniksa commented on src/inc/til/point.h:
```diff
@@ -3,403 +3,264 @@
-            point(TilMath::template cast<ptrdiff_t>(other.X), TilMath::template cast<ptrdiff_t>(other.Y))
+        constexpr bool operator==(const point rhs) const noexcept
         {
+            return __builtin_memcmp(this, &rhs, sizeof(rhs)) == 0;
```
Why the private `__builtin_memcmp` instead of the public `memcmp`?

miniksa replied:
(leave comment at least...)

lhecker replied:
I added the following comment to all 3 files:
```cpp
// `__builtin_memcmp` isn't an official standard, but it's the
// only way at the time of writing to get a constexpr `memcmp`.
```

See [`P1944R1`](http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2020/p1944r1.pdf), section III B.

miniksa replied:
Blah. OK. Thank you.

miniksa commented on src/renderer/dx/DxRenderer.cpp:
```diff
@@ -1674,13 +1679,13 @@ CATCH_RETURN()
 // - S_OK or relevant DirectX error
 [[nodiscard]] HRESULT DxEngine::PaintBufferLine(gsl::span<const Cluster> const clusters,
-                                                COORD const coord,
+                                                COORD coord,
```
Not sure why `const` disappeared. You're not modifying it as far as I can tell.

lhecker replied:
Much of this PR was sponsored by the Search & Replace Inc. (with regexp). This is probably one of the many oopsies I forgot to revert. 😅

miniksa replied:
Wuh oh.

miniksa commented on src/terminal/adapter/FontBuffer.cpp:
```diff
@@ -222,9 +222,9 @@ gsl::span<const uint16_t> FontBuffer::GetBitPattern() const noexcept
+til::size FontBuffer::GetCellSize() const noexcept
 {
-    return { _fullWidth, _fullHeight };
+    return { gsl::narrow_cast<til::CoordType>(_fullWidth), gsl::narrow_cast<til::CoordType>(_fullHeight) };
```
This is weird... having to do the narrow_cast on the outside by digging into `til::CoordType`... we can't just accept this into the constructor and narrow it there especially if we're going to do `narrow_cast` (and not throw on error)?

miniksa replied:
Oh, alright. I guess in context of the "whole story" then... it's fine. I'll let it pass. I keep remembering that you told me that all of this becomes less and less a problem the further we get to having actually unified types.

miniksa commented on src/til/ut_til/SPSCTests.cpp:
```diff
@@ -4,6 +4,8 @@
 #include "precomp.h"
 #include "WexTestClass.h"
 
+#include <til/spsc.h>
```
LOL. Oops.

miniksa commented on src/inc/til/rect.h:
```diff
@@ -0,0 +1,856 @@
+            return *this;
+        }
+
+        // scale_up will scale the entire rect up by the size factor
+        // This includes moving the origin.
+        constexpr rect scale_up(const size size) const
+        {
+            const auto topLeft = point{ left, top } * size;
+            const auto bottomRight = point{ right, bottom } * size;
+            return rect{ topLeft, bottomRight };
+        }
+
+        // scale_down will scale the entire rect down by the size factor,
+        // but rounds the bottom-right corner out.
+        // This includes moving the origin.
```
Make some ASCII art rectangle to match the above! :P

miniksa commented on src/inc/til/rect.h:
```diff
@@ -0,0 +1,856 @@
+        }
+#endif
+
+        std::wstring to_string() const
+        {
+            return wil::str_printf<std::wstring>(L"(L:%d, T:%d, R:%d, B:%d) [W:%d, H:%d]", left, top, right, bottom, width(), height());
+        }
```
Is this used outside of testing scenarios?

lhecker replied:
I renamed the file and this is the original code (almost everything in rect.h is unmodified!).
I'll add some ifdefs around this. 🙂

miniksa replied:
Oh sure. Feel free to ignore me then.

lhecker commented on src/cascadia/PublicTerminalCore/HwndTerminal.cpp:
```diff
@@ -582,19 +582,25 @@ try
+        const auto dx = cursorPosition.x - touchdownPoint.x;
+        const auto dy = cursorPosition.y - touchdownPoint.y;
+        const auto w = fontSize.width;
+        const auto distanceSquared = dx * dx + dy * dy;
+        const auto maxDistanceSquared = w * w / 16; // (w /4)^2
+
+        if (distanceSquared >= maxDistanceSquared)
```
No. Unless I made a weird mistake, the functionality is exactly the same. 🙂
The change is necessary since the old code wouldn't compile otherwise. Since a change was necessary I changed the distance check to a simpler standard distance check. It's rather unusual to use sqrt for that to be honest.
It's safer to just square the other side of the equation - no NANs & no explicit static_cast casts (which would've been necessary here).

lhecker commented on src/inc/til/size.h:
```diff
@@ -3,195 +3,105 @@
-            THROW_HR_IF(E_ABORT, !base::MakeCheckedNum(width).AssignIfValid(&_width));
-            THROW_HR_IF(E_ABORT, !base::MakeCheckedNum(height).AssignIfValid(&_height));
-        }
+        CoordType width = 0;
```
Yeah I agree. Do you have a suggestion how to improve this? Otherwise we'll probably have to rely on our IDEs for the moment. 🙁

miniksa replied:
No, we can deal.

lhecker commented on src/inc/til/point.h:
```diff
@@ -3,368 +3,229 @@
-            THROW_HR_IF(E_ABORT, !base::MakeCheckedNum(_x).AssignIfValid(&ret.x));
-            THROW_HR_IF(E_ABORT, !base::MakeCheckedNum(_y).AssignIfValid(&ret.y));
-            return ret;
+            return { x, y };
```
I am! It makes finding stuff with an editor so much easier! But in this case the type is just a line above, so I think that's fine. (And some of the later types are rather lengthy due to the namespace.) 🙂

lhecker commented on src/inc/til/bitmap.h:
```diff
@@ -76,11 +78,11 @@ namespace til // Terminal Implementation Library. Also: "Today I Learned"
-            const ptrdiff_t _end;
-            til::rectangle _run;
+            const til::rect _rc;
+            size_t _pos;
```
I do but it's a bit difficult to be thorough _and_ minimal in this PR. I found through experimentation that I get the smallest bitmap.h diff if I use `size_t` here. 🙂

miniksa replied:
Good enough.

carlos-zamora commented on src/interactivity/win32/ut_interactivity_win32/GeneratedUiaTextRangeMovementTests.g.cpp:
```diff
@@ -28,7 +28,7 @@ constexpr auto docEndM1L{ point_offset_by_line(docEnd, bufferSize, -1) };
 constexpr auto docEndM5C{ point_offset_by_char(docEnd, bufferSize, -5) };
 constexpr auto docEndM5L{ point_offset_by_line(docEnd, bufferSize, -5) };
 constexpr auto docEndP1C{ point_offset_by_char(docEnd, bufferSize, 1) };
-constexpr til::point lastCharPosLeft{ bufferSize.left(), lastCharPos.y() };
```
oh shoot, yeah. Could you just update `GenerateTests.ps1` then run the script again? Just to make sure we have the generated version committed.

lhecker commented:
Sorry for the force-push! I wanted to ensure that people who have reviewed before only need to review a single new commit.
It has the following changes:
* Address feedback you all brought up
* Pass point/size by value (fits into a single register) and rect by reference (doesn't fit) (until MSVC gets smart about this)
* Fix compilation of WddmCon/Bgfx engines (hopefully - I didn't test it)
* Simplify some of the code changes

miniksa reviewed (approved):
Let's do this!
